### ISSUE TITLE
Automate documenting workflows

### ISSUE THREAD

**Issue by KhudaDad414** [asyncapi/.github#104]
#### Reason/Context
Since we are using workflows in our repos a lot and the number of workflows is growing day by day, It is becoming harder to keep track of what workflows are doing and having a big picture of the automation that we do for each repo. for example, what will happen if I open a new issue?

#### Description
We need a way to document the workflows that we use. I have collected this [info about the workflows](https://docs.google.com/spreadsheets/d/1VnO0qqh0aSBN8rCue81vkNzH_a1zFpcwCDU6Vrq-53E/edit?usp=sharing) that we use in the `website` repo. that table can answer some questions about the workflows but it is definitely not the best possible option.
#### How could it be implemented/designed?
- First of all, we need a template for the documentation. [this table](https://docs.google.com/spreadsheets/d/1VnO0qqh0aSBN8rCue81vkNzH_a1zFpcwCDU6Vrq-53E/edit?usp=sharing) can show what event triggers what actions in general but I think we should document what steps each workflow takes as well. I am not exactly sure, maybe @alequetzalli has some idea on how we can document workflows.
- We will need a library that can be used as a workflow to generate the documentation on workflows just like [jsdoc2md](https://github.com/jsdoc2md) does for jsdoc.

**Comment by derberg** [comment]
Yes! that would help. I personally do not need detailed info about steps. For me most important is to have:
- list of workflows with filename -> Name -> Description (we need to figure out something, as there is no such field in GH Action workflow file)
- the table that you already create, so lovely overview of all the events to which our workflow reacts to


and then yes, we do not even need a new workflow as a workflow to update files in repo is there, it is just fixed to trigger only when `docs: ` PR is merged, we will just need to extend it with `ci: ` PRs

**Comment by KhudaDad414** [comment]
@derberg 
this is the structure that the action that I am working on generates. would love to have some feedback. :)
### When Pull Request Target: 
| opened | ready_for_review, reopened | synchronize, edited | labeled, unlocked, unlabeled | 
| --- | --- | --- | --- | 
| <ul><li>[Automerge release bump PR](#automerge-release-bump-pr)</li><li>[Notify slack](#notify-slack)</li><li>[Lint PR title](#lint-pr-title)</li><li>[Welcome first time contributors](#welcome-first-time-contributors)</li> </ul> |<ul><li>[Automerge release bump PR](#automerge-release-bump-pr)</li><li>[Notify slack](#notify-slack)</li><li>[Lint PR title](#lint-pr-title)</li> </ul> |<ul><li>[Automerge release bump PR](#automerge-release-bump-pr)</li><li>[Lint PR title](#lint-pr-title)</li> </ul> |<ul><li>[Automerge release bump PR](#automerge-release-bump-pr)</li> </ul> |
### When Issues: 
| opened | reopened | deleted, unlabeled, labeled, closed | edited | 
| --- | --- | --- | --- | 
| <ul><li>[Deploy to Netlify](#deploy-to-netlify)</li><li>[Notify slack](#notify-slack)</li><li>[Sentiment Analysis](#sentiment-analysis)</li><li>[Welcome first time contributors](#welcome-first-time-contributors)</li> </ul> |<ul><li>[Deploy to Netlify](#deploy-to-netlify)</li><li>[Notify slack](#notify-slack)</li> </ul> |<ul><li>[Deploy to Netlify](#deploy-to-netlify)</li> </ul> |<ul><li>[Sentiment Analysis](#sentiment-analysis)</li> </ul> |
### When Push:
| master | {} | next, **-release |
| --- | --- | --- |
| <ul><li>[Bump package version in dependent repos - if Node project](#bump-package-version-in-dependent-repos---if-node-project)</li><li>[Release - if Node project](#release---if-node-project)</li> </ul> |<ul><li>[autoupdate](#autoupdate)</li> </ul> |<ul><li>[Release - if Node project](#release---if-node-project)</li> </ul> |
### When Pull Request:
| opened | reopened, ready_for_review, synchronize | edited |
| --- | --- | --- |
| <ul><li>[PR testing - if Go project](#pr-testing---if-go-project)</li><li>[PR testing - if Node project](#pr-testing---if-node-project)</li><li>[Sentiment Analysis](#sentiment-analysis)</li> </ul> |<ul><li>[PR testing - if Go project](#pr-testing---if-go-project)</li><li>[PR testing - if Node project](#pr-testing---if-node-project)</li> </ul> |<ul><li>[Sentiment Analysis](#sentiment-analysis)</li> </ul> |
### When Pull Request Review:
| submitted | edited |
| --- | --- |
| <ul><li>[Automerge release bump PR](#automerge-release-bump-pr)</li><li>[Sentiment Analysis](#sentiment-analysis)</li> </ul> |<ul><li>[Sentiment Analysis](#sentiment-analysis)</li> </ul> |
### Schedules:
| At 12:00 AM |
| --- |
| <ul><li>[Notify on failing automerge](#notify-on-failing-automerge)</li><li>[Manage stale issues and PRs](#manage-stale-issues-and-prs)</li> </ul> |
### When Release:
| published |
| --- |
| <ul><li>[Version bump - if Node.js project](#version-bump---if-node.js-project)</li><li>[Announce releases in different channels](#announce-releases-in-different-channels)</li> </ul> |
### When Discussion:
| created |
| --- |
| <ul><li>[Notify slack](#notify-slack)</li> </ul> |
### When Issue Comment:
| created, edited |
| --- |
| <ul><li>[Sentiment Analysis](#sentiment-analysis)</li> </ul> |
### When Pull Request Review Comment:
| created, edited |
| --- |
| <ul><li>[Sentiment Analysis](#sentiment-analysis)</li> </ul> |## Workflows
### [Automerge release bump PR](.github/workflows/automerge.yml)
A short description about the workflow will appear here...
### [Notify slack](.github/workflows/issues-prs-notifications.yml)
A short description about the workflow will appear here...
### [Lint PR title](.github/workflows/lint-pr-title.yml)
A short description about the workflow will appear here...
### [Welcome first time contributors](.github/workflows/welcome-first-time-contrib.yml)
A short description about the workflow will appear here...
### [Deploy to Netlify](.github/workflows/deploy.yml)
A short description about the workflow will appear here...
### [Sentiment Analysis](.github/workflows/sentiment-analysis.yml)
A short description about the workflow will appear here...
### [Bump package version in dependent repos - if Node project](.github/workflows/bump.yml)
A short description about the workflow will appear here...
### [Release - if Node project](.github/workflows/if-nodejs-release.yml)
A short description about the workflow will appear here...
### [autoupdate](.github/workflows/autoupdate.yml)
A short description about the workflow will appear here...
### [PR testing - if Go project](.github/workflows/if-go-pr-testing.yml)
A short description about the workflow will appear here...
### [PR testing - if Node project](.github/workflows/if-nodejs-pr-testing.yml)
A short description about the workflow will appear here...
### [Notify on failing automerge](.github/workflows/automerge-orphans.yml)
A short description about the workflow will appear here...
### [Manage stale issues and PRs](.github/workflows/stale-issues-prs.yml)
A short description about the workflow will appear here...
### [Version bump - if Node.js project](.github/workflows/if-nodejs-version-bump.yml)
A short description about the workflow will appear here...
### [Announce releases in different channels](.github/workflows/release-announcements.yml)
A short description about the workflow will appear here...

**Comment by KhudaDad414** [comment]
just published [workflows-to-markdown](https://github.com/KhudaDad414/workflows-to-markdown) to help generate the markdown file easily (by adding `workflows2md` to `generate:assets` script).
we just have to decide which info is going to be included in the file and how exactly.

**Comment by derberg** [comment]
> what if we consider the first comment after workflow name as a short description?

yeah, I think there is no other option.

> Pull Request Target with Pull Request

they have a different scope of working, good to have it separated

regarding the layout of the information, I have to be honest and say I liked one table as you showed it in initial example. So first column event names, first-row event types. I think the list of workflows with their descriptions could also be a table. Now it is a bit overwhelming to digest I think

**Comment by derberg** [comment]
> thanks for the feedback, mate

no worries, happy to help, it is a cool tool

> updated the comment with the new design

hmmm I don't see any updates

> anything else that could make it better? what do you think about the CLI tool? do you prefer a GitHub action or CLI is better?

CLI is enough, I see you have it there already. We do not need dedicated GH action for our use case, as it will be just another `npm script` that will have to be invoked in `generate:assets` script as we do it [here for example](https://github.com/asyncapi/parser-js/blob/master/package.json#L25). But for your library would be cool to have in readme instruction for anyone how CLI can be run in GitHub Actions, like a copy paste template.

we only need to figure out what about repos that are not `node` 😅

**Comment by derberg** [comment]
cool, see a table now:
- Regarding "When Pull Request Target" as it is technical, can't it be `pull_request_target`, I mean code, without "When". Same about event "types", could be with markdown "code"

**Comment by KhudaDad414** [comment]
> we only need to figure out what about repos that are not `node`.
I am currently working on an action that non-`node` projects can add to their workflows.

**Comment by derberg** [comment]
you mean dedicated action, or a workflow that just calls CLI? like we do it with semantic release in not-node projects? so you just setup node env in one step and then just run `npx YOUR_CLI`.

although dedicated action is nice as it will expose your library to wider audience I think

**Comment by KhudaDad414** [comment]
In the past few days, I have been testing the action with all of the workflows across the organization. https://github.com/marketplace/actions/generate-workflows-map.
I think it is ready.

**Comment by derberg** [comment]
what is the final output format? btw you do not show any example, not in a CLI nor in the action. I think it is good to show it as a screen shot or just as part of markdown

**Comment by KhudaDad414** [comment]
usage and examples added for both repos. 
https://github.com/KhudaDad414/workflows-to-markdown
https://github.com/KhudaDad414/workflows2md-action

**Comment by derberg** [comment]
Looks great! so where are we putting it? I guess these MD files should land in `.github/workflows`?

**Comment by KhudaDad414** [comment]
@derberg For us, `.github/workflows` seems to be the best place to put it. by default I am not putting it there since workflow would need `workflows:write` permission. I will open a PR, adding it to this repo.

**Comment by derberg** [comment]
ok, keep in mind that in our case we are not pushing directly to upstream but we go through PR flow

**Comment by KhudaDad414** [comment]
@derberg I know we talked about using the tool in `generate:assets` script and not in a separate workflow but by using a workflow we can manage it for all repos (node and non-node) from here. we just need to run the workflow whenever repo merges a `ci: ...` PR. what do you think?

**Comment by derberg** [comment]
I'm only afraid of additional load we cause on CI, we are getting into dangerous lands.

So now we update global workflow changes in every repo with https://github.com/asyncapi/community/pull/229, and this would trigger another workflow that would update docs.

Both ☝🏼 workflows trigger tests and load on CI, and basically will make it stuck for an hour or 2. We somehow need to find a solution on how to skip CI on this kind of PRs.

We need to solve https://github.com/asyncapi/.github/issues/116 before we can add more workflows that we know will trigger load. Thoughts?
